In [1]:
import numpy as np
import pandas as pd
from pathlib import Path

In [2]:
# load the dataset
airlines = pd.read_csv('data/airlines.csv')
airports = pd.read_csv('data/airports.csv')
flights = pd.read_csv('data/flights.csv',low_memory=False)

#### Airlines data

In [3]:
# check first few rows
airlines.head()

IATA_CODE                 AIRLINE
0        UA   United Air Lines Inc.
1        AA  American Airlines Inc.
2        US         US Airways Inc.
3        F9  Frontier Airlines Inc.
4        B6         JetBlue Airways

In [4]:
# check data types
airlines.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14 entries, 0 to 13
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   IATA_CODE  14 non-null     object
 1   AIRLINE    14 non-null     object
dtypes: object(2)
memory usage: 352.0+ bytes


In [5]:
# check airlines
airlines['AIRLINE'].value_counts()

United Air Lines Inc.           1
American Airlines Inc.          1
US Airways Inc.                 1
Frontier Airlines Inc.          1
JetBlue Airways                 1
Skywest Airlines Inc.           1
Alaska Airlines Inc.            1
Spirit Air Lines                1
Southwest Airlines Co.          1
Delta Air Lines Inc.            1
Atlantic Southeast Airlines     1
Hawaiian Airlines Inc.          1
American Eagle Airlines Inc.    1
Virgin America                  1
Name: AIRLINE, dtype: int64

In [6]:
# check size
airlines.shape

(14, 2)

In [7]:
# check missing values
airlines.isna().sum()

IATA_CODE    0
AIRLINE      0
dtype: int64

No missing values

Airlines data is relatively small, and no need for further data cleaning

#### Airports data

In [8]:
# check first few rows
airports.head()

IATA_CODE                              AIRPORT         CITY STATE COUNTRY  \
0       ABE  Lehigh Valley International Airport    Allentown    PA     USA   
1       ABI             Abilene Regional Airport      Abilene    TX     USA   
2       ABQ    Albuquerque International Sunport  Albuquerque    NM     USA   
3       ABR            Aberdeen Regional Airport     Aberdeen    SD     USA   
4       ABY   Southwest Georgia Regional Airport       Albany    GA     USA   

   LATITUDE  LONGITUDE  
0  40.65236  -75.44040  
1  32.41132  -99.68190  
2  35.04022 -106.60919  
3  45.44906  -98.42183  
4  31.53552  -84.19447

In [9]:
# check sizes
airports.shape

(322, 7)

In [10]:
# check data types
airports.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 322 entries, 0 to 321
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   IATA_CODE  322 non-null    object 
 1   AIRPORT    322 non-null    object 
 2   CITY       322 non-null    object 
 3   STATE      322 non-null    object 
 4   COUNTRY    322 non-null    object 
 5   LATITUDE   319 non-null    float64
 6   LONGITUDE  319 non-null    float64
dtypes: float64(2), object(5)
memory usage: 17.7+ KB


In [11]:
# check missing values
airports.isna().sum()

IATA_CODE    0
AIRPORT      0
CITY         0
STATE        0
COUNTRY      0
LATITUDE     3
LONGITUDE    3
dtype: int64

There are 3 missing values for 'LATITUDE' and 'LONGITUDE'

In [12]:
# drop rows with missing values
airports = airports.dropna()

In [13]:
# check missing values again
airports.isna().sum()

IATA_CODE    0
AIRPORT      0
CITY         0
STATE        0
COUNTRY      0
LATITUDE     0
LONGITUDE    0
dtype: int64

No missing values now

#### Flights data

In [14]:
# check size
flights.shape

(5819079, 31)

It contains 5819079 rows and 31 columns

In [15]:
# check data type
flights.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5819079 entries, 0 to 5819078
Data columns (total 31 columns):
 #   Column               Dtype  
---  ------               -----  
 0   YEAR                 int64  
 1   MONTH                int64  
 2   DAY                  int64  
 3   DAY_OF_WEEK          int64  
 4   AIRLINE              object 
 5   FLIGHT_NUMBER        int64  
 6   TAIL_NUMBER          object 
 7   ORIGIN_AIRPORT       object 
 8   DESTINATION_AIRPORT  object 
 9   SCHEDULED_DEPARTURE  int64  
 10  DEPARTURE_TIME       float64
 11  DEPARTURE_DELAY      float64
 12  TAXI_OUT             float64
 13  WHEELS_OFF           float64
 14  SCHEDULED_TIME       float64
 15  ELAPSED_TIME         float64
 16  AIR_TIME             float64
 17  DISTANCE             int64  
 18  WHEELS_ON            float64
 19  TAXI_IN              float64
 20  SCHEDULED_ARRIVAL    int64  
 21  ARRIVAL_TIME         float64
 22  ARRIVAL_DELAY        float64
 23  DIVERTED             int64  
 24

In [16]:
# Check basic stats
flights.describe()

YEAR         MONTH           DAY   DAY_OF_WEEK  FLIGHT_NUMBER  \
count  5819079.0  5.819079e+06  5.819079e+06  5.819079e+06   5.819079e+06   
mean      2015.0  6.524085e+00  1.570459e+01  3.926941e+00   2.173093e+03   
std          0.0  3.405137e+00  8.783425e+00  1.988845e+00   1.757064e+03   
min       2015.0  1.000000e+00  1.000000e+00  1.000000e+00   1.000000e+00   
25%       2015.0  4.000000e+00  8.000000e+00  2.000000e+00   7.300000e+02   
50%       2015.0  7.000000e+00  1.600000e+01  4.000000e+00   1.690000e+03   
75%       2015.0  9.000000e+00  2.300000e+01  6.000000e+00   3.230000e+03   
max       2015.0  1.200000e+01  3.100000e+01  7.000000e+00   9.855000e+03   

       SCHEDULED_DEPARTURE  DEPARTURE_TIME  DEPARTURE_DELAY      TAXI_OUT  \
count         5.819079e+06    5.732926e+06     5.732926e+06  5.730032e+06   
mean          1.329602e+03    1.335204e+03     9.370158e+00  1.607166e+01   
std           4.837518e+02    4.964233e+02     3.708094e+01  8.895574e+00   
min           1.000000e+00    1.000000e+00    -8.200000e+01  1.000000e+00   
25%           9.170000e+02    9.210000e+02    -5.000000e+00  1.100000e+01   
50%           1.325000e+03    1.330000e+03    -2.000000e+00  1.400000e+01   
75%           1.730000e+03    1.740000e+03     7.000000e+00  1.900000e+01   
max           2.359000e+03    2.400000e+03     1.988000e+03  2.250000e+02   

         WHEELS_OFF  ...  SCHEDULED_ARRIVAL  ARRIVAL_TIME  ARRIVAL_DELAY  \
count  5.730032e+06  ...       5.819079e+06  5.726566e+06   5.714008e+06   
mean   1.357171e+03  ...       1.493808e+03  1.476491e+03   4.407057e+00   
std    4.980094e+02  ...       5.071647e+02  5.263197e+02   3.927130e+01   
min    1.000000e+00  ...       1.000000e+00  1.000000e+00  -8.700000e+01   
25%    9.350000e+02  ...       1.110000e+03  1.059000e+03  -1.300000e+01   
50%    1.343000e+03  ...       1.520000e+03  1.512000e+03  -5.000000e+00   
75%    1.754000e+03  ...       1.918000e+03  1.917000e+03   8.000000e+00   
max    2.400000e+03  ...       2.400000e+03  2.400000e+03   1.971000e+03   

           DIVERTED     CANCELLED  AIR_SYSTEM_DELAY  SECURITY_DELAY  \
count  5.819079e+06  5.819079e+06      1.063439e+06    1.063439e+06   
mean   2.609863e-03  1.544643e-02      1.348057e+01    7.615387e-02   
std    5.102012e-02  1.233201e-01      2.800368e+01    2.143460e+00   
min    0.000000e+00  0.000000e+00      0.000000e+00    0.000000e+00   
25%    0.000000e+00  0.000000e+00      0.000000e+00    0.000000e+00   
50%    0.000000e+00  0.000000e+00      2.000000e+00    0.000000e+00   
75%    0.000000e+00  0.000000e+00      1.800000e+01    0.000000e+00   
max    1.000000e+00  1.000000e+00      1.134000e+03    5.730000e+02   

       AIRLINE_DELAY  LATE_AIRCRAFT_DELAY  WEATHER_DELAY  
count   1.063439e+06         1.063439e+06   1.063439e+06  
mean    1.896955e+01         2.347284e+01   2.915290e+00  
std     4.816164e+01         4.319702e+01   2.043334e+01  
min     0.000000e+00         0.000000e+00   0.000000e+00  
25%     0.000000e+00         0.000000e+00   0.000000e+00  
50%     2.000000e+00         3.000000e+00   0.000000e+00  
75%     1.900000e+01         2.900000e+01   0.000000e+00  
max     1.971000e+03         1.331000e+03   1.211000e+03  

[8 rows x 26 columns]

In [17]:
# Check first rows
flights.head()

YEAR  MONTH  DAY  DAY_OF_WEEK AIRLINE  FLIGHT_NUMBER TAIL_NUMBER  \
0  2015      1    1            4      AS             98      N407AS   
1  2015      1    1            4      AA           2336      N3KUAA   
2  2015      1    1            4      US            840      N171US   
3  2015      1    1            4      AA            258      N3HYAA   
4  2015      1    1            4      AS            135      N527AS   

  ORIGIN_AIRPORT DESTINATION_AIRPORT  SCHEDULED_DEPARTURE  ...  ARRIVAL_TIME  \
0            ANC                 SEA                    5  ...         408.0   
1            LAX                 PBI                   10  ...         741.0   
2            SFO                 CLT                   20  ...         811.0   
3            LAX                 MIA                   20  ...         756.0   
4            SEA                 ANC                   25  ...         259.0   

   ARRIVAL_DELAY  DIVERTED  CANCELLED  CANCELLATION_REASON  AIR_SYSTEM_DELAY  \
0          -22.0         0          0                  NaN               NaN   
1           -9.0         0          0                  NaN               NaN   
2            5.0         0          0                  NaN               NaN   
3           -9.0         0          0                  NaN               NaN   
4          -21.0         0          0                  NaN               NaN   

   SECURITY_DELAY  AIRLINE_DELAY  LATE_AIRCRAFT_DELAY  WEATHER_DELAY  
0             NaN            NaN                  NaN            NaN  
1             NaN            NaN                  NaN            NaN  
2             NaN            NaN                  NaN            NaN  
3             NaN            NaN                  NaN            NaN  
4             NaN            NaN                  NaN            NaN  

[5 rows x 31 columns]

In [18]:
# Check last few rows
flights.tail()

YEAR  MONTH  DAY  DAY_OF_WEEK AIRLINE  FLIGHT_NUMBER TAIL_NUMBER  \
5819074  2015     12   31            4      B6            688      N657JB   
5819075  2015     12   31            4      B6            745      N828JB   
5819076  2015     12   31            4      B6           1503      N913JB   
5819077  2015     12   31            4      B6            333      N527JB   
5819078  2015     12   31            4      B6            839      N534JB   

        ORIGIN_AIRPORT DESTINATION_AIRPORT  SCHEDULED_DEPARTURE  ...  \
5819074            LAX                 BOS                 2359  ...   
5819075            JFK                 PSE                 2359  ...   
5819076            JFK                 SJU                 2359  ...   
5819077            MCO                 SJU                 2359  ...   
5819078            JFK                 BQN                 2359  ...   

         ARRIVAL_TIME  ARRIVAL_DELAY  DIVERTED  CANCELLED  \
5819074         753.0          -26.0         0          0   
5819075         430.0          -16.0         0          0   
5819076         432.0           -8.0         0          0   
5819077         330.0          -10.0         0          0   
5819078         442.0            2.0         0          0   

         CANCELLATION_REASON  AIR_SYSTEM_DELAY  SECURITY_DELAY  AIRLINE_DELAY  \
5819074                  NaN               NaN             NaN            NaN   
5819075                  NaN               NaN             NaN            NaN   
5819076                  NaN               NaN             NaN            NaN   
5819077                  NaN               NaN             NaN            NaN   
5819078                  NaN               NaN             NaN            NaN   

         LATE_AIRCRAFT_DELAY  WEATHER_DELAY  
5819074                  NaN            NaN  
5819075                  NaN            NaN  
5819076                  NaN            NaN  
5819077                  NaN            NaN  
5819078                  NaN            NaN  

[5 rows x 31 columns]

In [19]:
# Check how many missing values
no_null = flights.isnull().sum()
no_null_ordered = no_null.sort_values(ascending=False)
no_null_ordered

CANCELLATION_REASON    5729195
WEATHER_DELAY          4755640
LATE_AIRCRAFT_DELAY    4755640
AIRLINE_DELAY          4755640
SECURITY_DELAY         4755640
AIR_SYSTEM_DELAY       4755640
AIR_TIME                105071
ARRIVAL_DELAY           105071
ELAPSED_TIME            105071
WHEELS_ON                92513
TAXI_IN                  92513
ARRIVAL_TIME             92513
TAXI_OUT                 89047
WHEELS_OFF               89047
DEPARTURE_DELAY          86153
DEPARTURE_TIME           86153
TAIL_NUMBER              14721
SCHEDULED_TIME               6
SCHEDULED_DEPARTURE          0
CANCELLED                    0
DAY                          0
DAY_OF_WEEK                  0
AIRLINE                      0
FLIGHT_NUMBER                0
SCHEDULED_ARRIVAL            0
DIVERTED                     0
ORIGIN_AIRPORT               0
DISTANCE                     0
DESTINATION_AIRPORT          0
MONTH                        0
YEAR                         0
dtype: int64

In [20]:
# Check % of missing values
percent_null = flights.isnull().sum() / len(flights) * 100
percent_null_ordered = percent_null.sort_values(ascending=False)
percent_null_ordered

CANCELLATION_REASON    98.455357
WEATHER_DELAY          81.724960
LATE_AIRCRAFT_DELAY    81.724960
AIRLINE_DELAY          81.724960
SECURITY_DELAY         81.724960
AIR_SYSTEM_DELAY       81.724960
AIR_TIME                1.805629
ARRIVAL_DELAY           1.805629
ELAPSED_TIME            1.805629
WHEELS_ON               1.589822
TAXI_IN                 1.589822
ARRIVAL_TIME            1.589822
TAXI_OUT                1.530259
WHEELS_OFF              1.530259
DEPARTURE_DELAY         1.480526
DEPARTURE_TIME          1.480526
TAIL_NUMBER             0.252978
SCHEDULED_TIME          0.000103
SCHEDULED_DEPARTURE     0.000000
CANCELLED               0.000000
DAY                     0.000000
DAY_OF_WEEK             0.000000
AIRLINE                 0.000000
FLIGHT_NUMBER           0.000000
SCHEDULED_ARRIVAL       0.000000
DIVERTED                0.000000
ORIGIN_AIRPORT          0.000000
DISTANCE                0.000000
DESTINATION_AIRPORT     0.000000
MONTH                   0.000000
YEAR      

We can see a large proportion of missing values of the first 6 columns

In [21]:
# Drop 'ARRIVAL_TIME' and 'SCHEDULED_ARRIVAL'
# Because 'ARRIVAL_DELAY' summarizes these 2 columns already
flights = flights.drop(['ARRIVAL_TIME','SCHEDULED_ARRIVAL'], axis=1)

In [22]:
# Similarly, we dont need 'SCHEDULED_DEPARTURE', 'DEPARTURE_TIME', 'TAXI_OUT', 'TAXI_IN', 'WHEELS_OFF','WHEELS_ON','AIR_TIME','SCHEDULED_TIME','ELAPSED_TIME'
flights = flights.drop(['SCHEDULED_DEPARTURE','DEPARTURE_TIME','TAXI_OUT',
                       'TAXI_IN','WHEELS_OFF','WHEELS_ON',
                       'AIR_TIME','SCHEDULED_TIME','ELAPSED_TIME'], axis=1)

In [23]:
# view columns left
flights.columns.sort_values()

Index(['AIRLINE', 'AIRLINE_DELAY', 'AIR_SYSTEM_DELAY', 'ARRIVAL_DELAY',
       'CANCELLATION_REASON', 'CANCELLED', 'DAY', 'DAY_OF_WEEK',
       'DEPARTURE_DELAY', 'DESTINATION_AIRPORT', 'DISTANCE', 'DIVERTED',
       'FLIGHT_NUMBER', 'LATE_AIRCRAFT_DELAY', 'MONTH', 'ORIGIN_AIRPORT',
       'SECURITY_DELAY', 'TAIL_NUMBER', 'WEATHER_DELAY', 'YEAR'],
      dtype='object')

In [24]:
# check missing values again
flights.isna().sum().sort_values(ascending=False)

CANCELLATION_REASON    5729195
WEATHER_DELAY          4755640
LATE_AIRCRAFT_DELAY    4755640
AIRLINE_DELAY          4755640
SECURITY_DELAY         4755640
AIR_SYSTEM_DELAY       4755640
ARRIVAL_DELAY           105071
DEPARTURE_DELAY          86153
TAIL_NUMBER              14721
CANCELLED                    0
DIVERTED                     0
YEAR                         0
MONTH                        0
DESTINATION_AIRPORT          0
ORIGIN_AIRPORT               0
FLIGHT_NUMBER                0
AIRLINE                      0
DAY_OF_WEEK                  0
DAY                          0
DISTANCE                     0
dtype: int64

In [25]:
# check 'TAIL_NUMBER' (aircraft identifier)
flights['TAIL_NUMBER'].value_counts()

N480HA    3768
N488HA    3723
N484HA    3723
N493HA    3585
N478HA    3577
          ... 
N175UA       1
N852NW       1
N7LBAA       1
N7LEAA       1
N180UA       1
Name: TAIL_NUMBER, Length: 4897, dtype: int64

In [26]:
# drop missing values under 'TAIL_NUMBER'
flights = flights.dropna(subset=['TAIL_NUMBER'])

In [27]:
# check size of new data
flights.shape

(5804358, 20)

The new flights data has 5804358 rows and 20 columns

#### Merge 3 datasets

In [28]:
# Inner join flights and airlines data
flights = flights.merge(airlines, left_on='AIRLINE',right_on='IATA_CODE',how='inner')

In [29]:
flights.head()

YEAR  MONTH  DAY  DAY_OF_WEEK AIRLINE_x  FLIGHT_NUMBER TAIL_NUMBER  \
0  2015      1    1            4        AS             98      N407AS   
1  2015      1    1            4        AS            135      N527AS   
2  2015      1    1            4        AS            108      N309AS   
3  2015      1    1            4        AS            122      N413AS   
4  2015      1    1            4        AS            130      N457AS   

  ORIGIN_AIRPORT DESTINATION_AIRPORT  DEPARTURE_DELAY  ...  DIVERTED  \
0            ANC                 SEA            -11.0  ...         0   
1            SEA                 ANC             -1.0  ...         0   
2            ANC                 SEA             -4.0  ...         0   
3            ANC                 PDX             -4.0  ...         0   
4            FAI                 SEA             -8.0  ...         0   

   CANCELLED  CANCELLATION_REASON  AIR_SYSTEM_DELAY SECURITY_DELAY  \
0          0                  NaN               NaN            NaN   
1          0                  NaN               NaN            NaN   
2          0                  NaN               NaN            NaN   
3          0                  NaN               NaN            NaN   
4          0                  NaN               NaN            NaN   

   AIRLINE_DELAY  LATE_AIRCRAFT_DELAY  WEATHER_DELAY  IATA_CODE  \
0            NaN                  NaN            NaN         AS   
1            NaN                  NaN            NaN         AS   
2            NaN                  NaN            NaN         AS   
3            NaN                  NaN            NaN         AS   
4            NaN                  NaN            NaN         AS   

              AIRLINE_y  
0  Alaska Airlines Inc.  
1  Alaska Airlines Inc.  
2  Alaska Airlines Inc.  
3  Alaska Airlines Inc.  
4  Alaska Airlines Inc.  

[5 rows x 22 columns]

In [30]:
# Remove redundant columns
flights = flights.drop(['AIRLINE_x','IATA_CODE'], axis=1)

In [31]:
# Inner join flights and airports data
flights = flights.merge(airports, left_on='ORIGIN_AIRPORT',right_on='IATA_CODE',how='inner')
flights = flights.merge(airports, left_on='DESTINATION_AIRPORT',right_on='IATA_CODE',how='inner')

In [32]:
# check new columns
flights.columns.sort_values()

Index(['AIRLINE_DELAY', 'AIRLINE_y', 'AIRPORT_x', 'AIRPORT_y',
       'AIR_SYSTEM_DELAY', 'ARRIVAL_DELAY', 'CANCELLATION_REASON', 'CANCELLED',
       'CITY_x', 'CITY_y', 'COUNTRY_x', 'COUNTRY_y', 'DAY', 'DAY_OF_WEEK',
       'DEPARTURE_DELAY', 'DESTINATION_AIRPORT', 'DISTANCE', 'DIVERTED',
       'FLIGHT_NUMBER', 'IATA_CODE_x', 'IATA_CODE_y', 'LATE_AIRCRAFT_DELAY',
       'LATITUDE_x', 'LATITUDE_y', 'LONGITUDE_x', 'LONGITUDE_y', 'MONTH',
       'ORIGIN_AIRPORT', 'SECURITY_DELAY', 'STATE_x', 'STATE_y', 'TAIL_NUMBER',
       'WEATHER_DELAY', 'YEAR'],
      dtype='object')

In [33]:
# check new data
flights.head()

YEAR  MONTH  DAY  DAY_OF_WEEK  FLIGHT_NUMBER TAIL_NUMBER ORIGIN_AIRPORT  \
0  2015      1    1            4             98      N407AS            ANC   
1  2015      1    1            4            108      N309AS            ANC   
2  2015      1    1            4            136      N431AS            ANC   
3  2015      1    1            4            134      N464AS            ANC   
4  2015      1    1            4            114      N303AS            ANC   

  DESTINATION_AIRPORT  DEPARTURE_DELAY  DISTANCE  ...  COUNTRY_x  LATITUDE_x  \
0                 SEA            -11.0      1448  ...        USA    61.17432   
1                 SEA             -4.0      1448  ...        USA    61.17432   
2                 SEA              NaN      1448  ...        USA    61.17432   
3                 SEA            -15.0      1448  ...        USA    61.17432   
4                 SEA            -11.0      1448  ...        USA    61.17432   

   LONGITUDE_x IATA_CODE_y                             AIRPORT_y   CITY_y  \
0   -149.99619         SEA  Seattle-Tacoma International Airport  Seattle   
1   -149.99619         SEA  Seattle-Tacoma International Airport  Seattle   
2   -149.99619         SEA  Seattle-Tacoma International Airport  Seattle   
3   -149.99619         SEA  Seattle-Tacoma International Airport  Seattle   
4   -149.99619         SEA  Seattle-Tacoma International Airport  Seattle   

   STATE_y  COUNTRY_y  LATITUDE_y LONGITUDE_y  
0       WA        USA    47.44898  -122.30931  
1       WA        USA    47.44898  -122.30931  
2       WA        USA    47.44898  -122.30931  
3       WA        USA    47.44898  -122.30931  
4       WA        USA    47.44898  -122.30931  

[5 rows x 34 columns]

In [34]:
# check if some columns are duplicates
print(flights['COUNTRY_x'].equals(flights['COUNTRY_y']))
print(flights['STATE_x'].equals(flights['STATE_y']))
print(flights['CITY_x'].equals(flights['CITY_y']))
print(flights['AIRPORT_x'].equals(flights['AIRPORT_y']))
print(flights['LATITUDE_x'].equals(flights['LATITUDE_y']))
print(flights['LONGITUDE_x'].equals(flights['LONGITUDE_y']))
print(flights['DESTINATION_AIRPORT'].equals(flights['IATA_CODE_y']))
print(flights['ORIGIN_AIRPORT'].equals(flights['IATA_CODE_x']))

True
False
False
False
False
False
True
True


In [35]:
# Remove redundant columns
flights = flights.drop(['COUNTRY_x','COUNTRY_y',
                       'STATE_x','STATE_y',
                       'IATA_CODE_x','IATA_CODE_y'], axis=1)

In [36]:
# Rename columns
flights = flights.rename(columns={'CITY_x': 'org_city',
                                  'CITY_y': 'dest_city',
                                  'AIRPORT_x': 'org_airport',
                                  'AIRPORT_y': 'dest_airport',
                                  'AIRLINE_y': 'airline',
                                  'LATITUDE_x': 'org_latitude',
                                  'LATITUDE_y': 'dest_latitude',
                                  'LONGITUDE_x': 'org_longitude',
                                  'LONGITUDE_y': 'dest_longitude',
                                  'TAIL_NUMBER': 'airplane_number'})

In [37]:
# check new column names
flights.columns

Index(['YEAR', 'MONTH', 'DAY', 'DAY_OF_WEEK', 'FLIGHT_NUMBER',
       'airplane_number', 'ORIGIN_AIRPORT', 'DESTINATION_AIRPORT',
       'DEPARTURE_DELAY', 'DISTANCE', 'ARRIVAL_DELAY', 'DIVERTED', 'CANCELLED',
       'CANCELLATION_REASON', 'AIR_SYSTEM_DELAY', 'SECURITY_DELAY',
       'AIRLINE_DELAY', 'LATE_AIRCRAFT_DELAY', 'WEATHER_DELAY', 'airline',
       'org_airport', 'org_city', 'org_latitude', 'org_longitude',
       'dest_airport', 'dest_city', 'dest_latitude', 'dest_longitude'],
      dtype='object')

In [38]:
# make all column names lower case
flights.columns = flights.columns.str.lower()

In [39]:
# check new data
flights.head()

year  month  day  day_of_week  flight_number airplane_number  \
0  2015      1    1            4             98          N407AS   
1  2015      1    1            4            108          N309AS   
2  2015      1    1            4            136          N431AS   
3  2015      1    1            4            134          N464AS   
4  2015      1    1            4            114          N303AS   

  origin_airport destination_airport  departure_delay  distance  ...  \
0            ANC                 SEA            -11.0      1448  ...   
1            ANC                 SEA             -4.0      1448  ...   
2            ANC                 SEA              NaN      1448  ...   
3            ANC                 SEA            -15.0      1448  ...   
4            ANC                 SEA            -11.0      1448  ...   

   weather_delay               airline  \
0            NaN  Alaska Airlines Inc.   
1            NaN  Alaska Airlines Inc.   
2            NaN  Alaska Airlines Inc.   
3            NaN  Alaska Airlines Inc.   
4            NaN  Alaska Airlines Inc.   

                                   org_airport   org_city  org_latitude  \
0  Ted Stevens Anchorage International Airport  Anchorage      61.17432   
1  Ted Stevens Anchorage International Airport  Anchorage      61.17432   
2  Ted Stevens Anchorage International Airport  Anchorage      61.17432   
3  Ted Stevens Anchorage International Airport  Anchorage      61.17432   
4  Ted Stevens Anchorage International Airport  Anchorage      61.17432   

   org_longitude                          dest_airport  dest_city  \
0     -149.99619  Seattle-Tacoma International Airport    Seattle   
1     -149.99619  Seattle-Tacoma International Airport    Seattle   
2     -149.99619  Seattle-Tacoma International Airport    Seattle   
3     -149.99619  Seattle-Tacoma International Airport    Seattle   
4     -149.99619  Seattle-Tacoma International Airport    Seattle   

   dest_latitude dest_longitude  
0       47.44898     -122.30931  
1       47.44898     -122.30931  
2       47.44898     -122.30931  
3       47.44898     -122.30931  
4       47.44898     -122.30931  

[5 rows x 28 columns]

In [40]:
# check new data size
flights.shape

(5309350, 28)

In [41]:
# save the data into a new excel sheet
datapath = Path('/Users/bq/Desktop/Bootcamp/coding/Springboard/Capstone3/data/flights_dataWrangling2.csv')
flights.to_csv(datapath)